Fix old GCC ABI check in CMake package config

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72081

This PR fixes the libstdc++ ABI check in CMake package configuration file (i.e. `TorchConfig.cmake`) The `_GLIBCXX_USE_CXX11_ABI` flag is a property of `libstdc++`, not GNU compiler collection. In its current form C++ libraries built with Clang on Linux fail since the `torch` CMake target propagates `_GLIBCXX_USE_CXX11_ABI` only when used with gcc.

Differential Revision: [D33899849](https://our.internmc.facebook.com/intern/diff/D33899849/)

cbalioglu committed:
Fix old GCC ABI check in CMake package config

This PR fixes the libstdc++ ABI check in CMake package configuration file (i.e. `TorchConfig.cmake`) The `_GLIBCXX_USE_CXX11_ABI` flag is a property of `libstdc++`, not GNU compiler collection. In its current form C++ libraries built with Clang on Linux fail since the `torch` CMake target propagates `_GLIBCXX_USE_CXX11_ABI` only when used with gcc.

Differential Revision: [D33899849](https://our.internmc.facebook.com/intern/diff/D33899849/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Fix old GCC ABI check in CMake package config"


This PR fixes the libstdc++ ABI check in CMake package configuration file (i.e. `TorchConfig.cmake`) The `_GLIBCXX_USE_CXX11_ABI` flag is a property of `libstdc++`, not GNU compiler collection. In its current form C++ libraries built with Clang on Linux fail since the `torch` CMake target propagates `_GLIBCXX_USE_CXX11_ABI` only when used with gcc.

Differential Revision: [D33899849](https://our.internmc.facebook.com/intern/diff/D33899849/)

[ghstack-poisoned]

malfet commented on cmake/TorchConfig.cmake.in:
```diff
@@ -166,9 +166,11 @@ if(@USE_CUDA@)
+# When we build libtorch with the old libstdc++ ABI, dependent libraries must too.
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
   set(TORCH_CXX_FLAGS "-D_GLIBCXX_USE_CXX11_ABI=@GLIBCXX_USE_CXX11_ABI@")
+else()
+  set(TORCH_CXX_FLAGS "-D_GLIBCXX_USE_CXX11_ABI=0")
```
Can you please elaborate why this is needed? This define should not do anything on non-Linux platforms, should it?

cbalioglu replied:
It shouldn't be needed. Just trying to figure out why most of the builds have started to fail after this change with a `error: invalid token at start of a preprocessor expression`.

malfet replied:
You probably just need to rebase your PR against commit from `viable/strict` (there were several regressions this morning)

cbalioglu replied:
Ok, resolved the issue. Apparently caffe2's CMakeLists.txt was the culprit and had a similar check that caused GLIBCXX_USE_CXX11_ABI  variable to become empty. Everything is green now.

cbalioglu committed:
Update on "Fix old GCC ABI check in CMake package config"


This PR fixes the libstdc++ ABI check in CMake package configuration file (i.e. `TorchConfig.cmake`) The `_GLIBCXX_USE_CXX11_ABI` flag is a property of `libstdc++`, not GNU compiler collection. In its current form C++ libraries built with Clang on Linux fail since the `torch` CMake target propagates `_GLIBCXX_USE_CXX11_ABI` only when used with gcc.

Differential Revision: [D33899849](https://our.internmc.facebook.com/intern/diff/D33899849/)

[ghstack-poisoned]
